extra_data/reader.py

```diff
@@ -354,6 +356,15 @@ def initializer():
     def from_path(cls, path):
         files = [FileAccess(path)]
         return cls(files, ctx_closes=True)
+    
+    def _close_all_files(self):
+        # Close the files if this collection was created by opening them.
+        if self.ctx_closes:
+            for file in self.files:
+                file.close()
+    
+    def __del__(self):
+        self._close_all_files()
```

Comment:
While running the tests for PR #47, I saw a load of errors from `__del__` methods running at Python shutdown. I'm not exactly sure how to reproduce that, but I know such problems are fairly common with `__del__` methods. I'd really like to avoid `__del__` methods wherever possible, and let normal object cleanup happen.

I'll try to have a look into the code relating to this myself and see what I can do, so hold off on this for the moment.

Reply:
Unfortunately it is absolutely necessary here. Without it the files stays open in the file cache after destroying `Data Collection` instance until garbage collection which can be delayed indefinitely. (there is check on it in tests, and they fail without this `__del__`)

In opposite, the `__del__` just closes files (calls `FileAccess.close()` which just assigns `None` to the `__file` attribute). I don't think it can raise exception ever on shutdown time.